doc/man3/EVP_EncryptInit.pod

```diff
@@ -137,7 +137,10 @@ as a result the amount of data written may be anything from zero bytes
 to (inl + cipher_block_size - 1) so B<out> should contain sufficient
 room. The actual number of bytes written is placed in B<outl>. It also
 checks if B<in> and B<out> are partially overlapping, and if they are
-0 is returned to indicate failure.
+0 is returned to indicate failure. Fully overlapping B<in> and B<out>
+will not lead to this. B<out> may be NULL to indicate that B<in> is
+additional data that shouldn't be encrypted but still used to calculate
+the MAC.
```

Comment:
this does not exactly translate to `is_partially_overlapping(out + ctx->buf_len, in, cmpl)` ?

Reply:
I meant "fully overlapping" to denote `in == out`, but english is not my first language.

Would "This does not happen if `B<in>` and `B<out>` have the same value" be a better alternative?

Reply:
I was not involved when this code was introduced, and I am not 100% sure if it
is correct at all.

But if ctx->buf_len != 0 and ctx->buf_len < cmpl then the is_partially_overlapped
condition would fail if in == out,  but if would pass if in = out + ctx->buf_len.

Reply:
`is_partially_overlapping()` always returns 0 if `in == out`, because it has a short-circuit-logic to do  that

> `PTRDIFF_T diff = (PTRDIFF_T)ptr1-(PTRDIFF_T)ptr2;`
> [...]
> `int overlapped = ` [...] ` & (diff != 0) & ` [...] ` ;`
> `return overlapped;`

Reply:
Yes, but here it is called with is_partially_overlapping(out + ctx->buf_len, in, cmpl)

Reply:
Ah, right. My bad.

Also it seems I overlooked [another instance of `is_partially_overlapping()` getting called](/kas-luthor/openssl/blob/812b74ae8cd1b4a51c14a4d1358a646a0420fb64/crypto/evp/evp_enc.c#L307) (which ironically does call `is_partially_overlapping(out, in, cmpl)`)

And `out` being incremented [here](https://github.com/kas-luthor/openssl/blob/812b74ae8cd1b4a51c14a4d1358a646a0420fb64/crypto/evp/evp_enc.c#L353)

I'll fix those too once it's cleared up how fully overlapping behaves

Reply:
I have no idea what the code should do or what you're trying to say. Are you trying to say that in == out is supported? I'm not sure that's always that case.

Reply:
What I'm trying to say is that `EVP_R_PARTIALLY_OVERLAPPING` won't necessarily be raised if `out == in`, though as of now I'm waiting to see if @**dot-asm** can clarify the situation